Add GeneralPanel Edit submit

jessie-ross opened the pull request:
<!-- Describe your changes in detail. -->
**This PR adds submission to the GeneralPanel Edit Modal:**
- Adds `ExperimentsApi.patch`
- Uses `ExperimentsApi.patch` for the submit.
- Beefs up the Save button in the modal.

Currently fails integration tests and local as swagger is failing, it seems to be requiring the `name` property to be present?
<!-- Remember to include context: Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

Next and Final PR is calling the refresh on the experiment data.

## How has this been tested?

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->
<!-- Delete any bullet points that don't apply and add more details if needed. -->

- Automated tests that cover added/changed functionality
- Manual testing with production data (locally)

yanirs commented on api/ExperimentsApi.ts:
```diff
@@ -24,6 +25,39 @@ async function create(newExperiment: ExperimentFullNew) {
+    //       We will add special logic for this one...
+    'metricAssignments',
+  ]
+  if (blacklistedKeys.some((key) => keys.includes(key))) {
```
Do we need this to be a runtime check? Can't we define `experimentPatch` to be of type `Partial<Omit<ExperimentFull, 'experimentId'  | 'exposureEvents' | ...>>`?

jessie-ross replied:
Well runtime has some advantages over compile-time, generally these functions (`patch` and `create`) don't actually get well typed inputs as we are receiving from user data. I would argue in this case it is preferable to do this runtime.

Although it is arguable whether this is useful at all...

yanirs replied:
> Well runtime has some advantages over compile-time, generally these functions (`patch` and `create`) don't actually get well typed inputs as we are receiving from user data. I would argue in this case it is preferable to do this runtime.

Looking at how it's called (with the output of `_.pick()`), there is some type checking going on. Also, an advantage of using the `Partial<Omit<...>>` approach is that the allowed fields are explicitly declared in the function signature. 

> Although it is arguable whether this is useful at all...

Yeah, maybe drop it completely? The backend will reject any fields that can't be updated anyway, and the list of blacklisted keys isn't exhaustive. If we do end up supporting additional fields on the backend, it'll be less work here. :slightly_smiling_face:

jessie-ross commented:
Ok, finally got to the failing integration tests:

> Currently fails integration tests and local as swagger is failing, it seems to be requiring the name property to be present?

@yanirs it might be worth you experimenting with swagger as you have it set up.

yanirs commented:
> Ok, finally got to the failing integration tests:
> 
> > Currently fails integration tests and local as swagger is failing, it seems to be requiring the name property to be present?
> 
> @yanirs it might be worth you experimenting with swagger as you have it set up.

It was because I didn't specify 200 as a possible response in the mock. :man_facepalming:
Fixed on SwaggerHub and I'll push a PR to update the doc.

yanirs commented on __tests__/api/ExperimentsApi.test.ts:
```diff
@@ -173,6 +173,45 @@ describe('ExperimentsApi.ts module', () => {
+        ownerLogin: 'owner-nickname',
+      }
+      const returnedExperiment = await validationErrorDisplayer(
+        ExperimentsApi.patch(0, (rawNewExperiment as unknown) as Partial<ExperimentFull>),
```
Let's use a realistic experiment ID (> 0).

yanirs commented on components/GeneralPanel.test.tsx:
```diff
@@ -185,6 +192,12 @@ test('checks edit dialog does not allow end datetime changes with disabled exper
   const experiment = Fixtures.createExperimentFull({ status: Status.Disabled })
   const { container: _container } = render(<GeneralPanel experiment={experiment} />)
 
+  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+  // @ts-ignore
+  global.fetch = jest.fn().mockImplementationOnce(async () => ({
+    json: async () => experiment,
+  }))
+
```
This is unnecessary as this test doesn't submit the form.

```suggestion
```

jessie-ross replied:
It happens to be needed now as we have slightly different submit logic for disabled experiments.

yanirs commented on components/GeneralPanel.test.tsx:
```diff
@@ -163,6 +164,12 @@ test('opens, submits and cancels edit dialog with running experiment', async ()
 
+  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+  // @ts-ignore
+  global.fetch = jest.fn().mockImplementationOnce(async () => ({
```
As this is a unit test, it'd be better to mock `ExperimentsApi.patch()` and verify that it's called with the expected fields on save. Otherwise, this test isn't really testing much (the behaviour of `Save` is identical to that of `Cancel`).

jessie-ross replied:
Mocking `ExperimentsApi.patch()` was a good idea!

yanirs reviewed (changes requested):
Let's make the panel test verify that Save and Cancel do different things. :slightly_smiling_face: 

Also, I did a quick check [on Vercel](https://abacus-git-add-general-experiments-submit.a8c.vercel.app/experiments/2) and we shouldn't be validating disabled fields:

![Peek 2020-09-02 08-22](https://user-images.githubusercontent.com/3952615/91912515-a73b1b00-ecf6-11ea-8ebc-8240ff7bdb0f.gif)

yanirs commented on components/GeneralPanel.tsx:
```diff
@@ -102,11 +119,16 @@ function GeneralPanel({ experiment }: { experiment: ExperimentFull }) {
-    // TODO: Full submission
-    enqueueSnackbar('Experiment Updated!', { variant: 'success' })
-    setIsEditing(false)
+  const onSubmitEdit = async (formData: { experiment: unknown }) => {
```
I think we can use `{ experiment: typeof generalEditInitialExperiment }` here.

yanirs commented on components/GeneralPanel.tsx:
```diff
@@ -46,6 +48,21 @@ const useStyles = makeStyles((theme: Theme) =>
         color: theme.palette.text.hint,
       },
     },
+    submitContainer: {
```
Shall we move some of these styles to a higher level? We're losing out on CSS benefits by localising this stuff.

jessie-ross replied:
It might be worth extracting this into a `ProgressButtonContainer` component

jessie-ross commented:
Ok, I've fixed the issues you have mentioned, these in particular:
- Made sure `endDatetime` isn't validated when it isn't editable.
- Extracted out `LoadingButtonContainer`

yanirs reviewed (approved):
I think we can improve the testing a bit, but nothing for this PR, so let's :ship:
